src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.cs

```diff
@@ -827,7 +827,27 @@ private async ValueTask<Stream> EstablishSocksTunnel(HttpRequestMessage request,
             return stream;
         }
 
-        private CancellationTokenSource GetConnectTimeoutCancellationTokenSource() => new CancellationTokenSource(Settings._connectTimeout);
+        private CancellationTokenSource GetConnectTimeoutCancellationTokenSource<T>(HttpConnectionWaiter<T> waiter)
+            where T : HttpConnectionBase?
+        {
+            var cts = new CancellationTokenSource(Settings._connectTimeout);
+
+            lock (waiter)
+            {
+                waiter.ConnectionCancellationTokenSource = cts;
+
+                // The initiating request for this connection attempt may complete concurrently at any time.
+                // If it completed before we've set the CTS, CancelIfNecessary would no-op.
+                // Check it again now that we're holding the lock and ensure we always set a timeout.
+                if (waiter.Task.IsCompleted)
+                {
+                    waiter.CancelIfNecessary(this, requestCancelled: waiter.Task.IsCanceled);
```

Comment:
I'm guilty of defining the `CancelIfNecessary` method, but with #100614 I find the method name confusing now, since it doesn't cancel the waiter, only the injected connection. Shall we consider renaming it in a separate PR? Or add clarification comments to the method name in this PR?

Reply:
I'll rename it here for main, but keep it as-is in backport PRs.